yktracec/src/promote.rs

```diff
@@ -0,0 +1,76 @@
+//! Software constant recorder.
+//!
+//! This is used for promoting variables to constants when compiling a trace.
+
+use std::{cell::RefCell, collections::VecDeque};
+
+// FIXME: Try to move this into ykrt in the existing thread local.
+thread_local! {
+    /// The thread's value recorder.
+    ///
+    /// This has to be specific to a thread to avoid cross-talk from other threads which may be
+    /// executing functions with promoted arguments at the same time.
+    pub static VAL_REC: RefCell<ValueRecorder> = RefCell::new(ValueRecorder::default());
+}
+
+/// A value recorder observes and records constant values during tracing. The trace compiler
+/// queries this when building a trace to replace promoted variables with the observed constants.
+#[derive(Debug, Default)]
+pub struct ValueRecorder {
+    // `true` when recording promotions.
+    record_enable: bool,
+    // A vector of observed values of to-be-promoted variables, in the order their `yk_promote()`
+    // calls were encountered in the trace.
+    //
+    // FIXME: currently you may only promote pointer-sized integers.
+    pbuf: VecDeque<usize>,
+}
+
+impl ValueRecorder {
+    /// Enable (`record=true`) or disable (`record=false`) observations.
+    pub fn record_enable(&mut self, record: bool) {
+        debug_assert_ne!(self.record_enable, record);
+        self.record_enable = record;
+    }
+
+    /// Record the concrete value of a variable to be promoted.
+    pub fn push(&mut self, val: usize) {
+        // OPT: a quicker way than a boolean check probably exists.
```

Comment:
I don't really understand this comment, and to the extent I do understand it, I'm not sure it's useful ;)

Reply:
I think `self.record_enable` isn't a very good way to decide if to record or not. We discussed in the past some kind of indirect jump to skip, or not, recording.

Reply:
I'm not sure that is possible and, if it is, I don't think this comment will remind us, to be honest.